server-spi/src/main/java/org/keycloak/models/credential/RecoveryAuthnCodesCredentialModel.java

```diff
@@ -0,0 +1,106 @@
+import java.util.stream.IntStream;
+
+import org.keycloak.credential.CredentialMetadata;
+import org.keycloak.credential.CredentialModel;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodeRepresentation;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodesCredentialData;
+import org.keycloak.models.credential.dto.RecoveryAuthnCodesSecretData;
+import org.keycloak.models.utils.RecoveryAuthnCodesUtils;
+import org.keycloak.util.JsonSerialization;
+
+import java.io.IOException;
+import java.util.List;
+
+public class RecoveryAuthnCodesCredentialModel extends CredentialModel {
+
+    public static final String TYPE = "recovery-authn-codes";
+
+    public static final String RECOVERY_CODES_NUMBER_USED = "recovery-codes-number-used";
+    public static final String RECOVERY_CODES_NUMBER_REMAINING = "recovery-codes-number-remaining";
+    public static final String RECOVERY_CODES_GENERATE_NEW_CODES = "recovery-codes-generate-new-codes";
+
+    private final RecoveryAuthnCodesCredentialData credentialData;
+    private final RecoveryAuthnCodesSecretData secretData;
+    private final CredentialMetadata credentialMetadata = null;
+
+    private RecoveryAuthnCodesCredentialModel(RecoveryAuthnCodesCredentialData credentialData,
+            RecoveryAuthnCodesSecretData secretData) {
+        this.credentialData = credentialData;
+        this.secretData = secretData;
+    }
+
+    public RecoveryAuthnCodeRepresentation getNextRecoveryAuthnCode() {
+        return this.secretData.getCodes().get(0);
+    }
+
+    public boolean allCodesUsed() {
+        return this.secretData.getCodes().isEmpty();
+    }
+
+    public void removeRecoveryAuthnCode() {
+        try {
+            this.secretData.removeNextBackupCode();
+            this.credentialData.setRemainingCodes(this.secretData.getCodes().size());
+            this.setSecretData(JsonSerialization.writeValueAsString(this.secretData));
+            this.setCredentialData(JsonSerialization.writeValueAsString(this.credentialData));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static RecoveryAuthnCodesCredentialModel createFromValues(List<String> originalGeneratedCodes, long generatedAt,
+            String userLabel) {
+        RecoveryAuthnCodesSecretData secretData;
+        RecoveryAuthnCodesCredentialData credentialData;
+        RecoveryAuthnCodesCredentialModel model;
+
+        try {
+            List<RecoveryAuthnCodeRepresentation> recoveryCodes = IntStream.range(0, originalGeneratedCodes.size())
+                    .mapToObj(i -> new RecoveryAuthnCodeRepresentation(i + 1,
+                            RecoveryAuthnCodesUtils.hashRawCode(originalGeneratedCodes.get(i))))
+                    .collect(Collectors.toList());
+            secretData = new RecoveryAuthnCodesSecretData(recoveryCodes);
+            credentialData = new RecoveryAuthnCodesCredentialData(RecoveryAuthnCodesUtils.NUM_HASH_ITERATIONS,
```

Comment:
Some environments might have stronger requirements then this defaults.

Could we make the number of iterations and the hash mechanism configurable, e.g. via password policies?

Alternatively extract the call to 
```
RecoveryAuthnCodesCredentialModel.createFromValues(generatedCodes,
                generatedAtTime, generatedUserLabel);
``` in `RecoveryAuthnCodesAction` into a protected method to allow users to apply a custom mechanism.

Reply:
OK.